### ISSUE TITLE
imagePull from private repo is slow second time

### ISSUE THREAD

**Issue by infa-ddeore** [XenitAB/spegel#217]
imagepull second time from another node in same cluster should be fast but it took same time as of downloading from remote repo, looks like caching didnt work 

here are the logs from spegel pod for `xxx.yyy.io/app-image-k8:dev_123` image (masked repo and image details)

any pointers to look into what can be the issue?

````
{"level":"error","ts":1696247240.4278097,"caller":"gin@v0.0.9/logger.go:62","msg":"","path":"/v2/app-image-k8/manifests/dev_123","status":404,"method":"HEAD","latency":5.000973923,"ip":"10.14.130.153","handler":"mirror","error":"could not resolve mirror for key: xxx.yyy.io/app-image-k8:dev_123","stacktrace":"github.com/xenitab/pkg/gin.Logger.func1\n\t/go/pkg/mod/github.com/xenitab/pkg/gin@v0.0.9/logger.go:62\ngithub.com/gin-gonic/gin.(*Context).Next\n\t/go/pkg/mod/github.com/gin-gonic/gin@v1.9.1/context.go:174\ngithub.com/gin-gonic/gin.(*Engine).handleHTTPRequest\n\t/go/pkg/mod/github.com/gin-gonic/gin@v1.9.1/gin.go:620\ngithub.com/gin-gonic/gin.(*Engine).ServeHTTP\n\t/go/pkg/mod/github.com/gin-gonic/gin@v1.9.1/gin.go:576\nnet/http.serverHandler.ServeHTTP\n\t/usr/local/go/src/net/http/server.go:2936\nnet/http.(*conn).serve\n\t/usr/local/go/src/net/http/server.go:1995"}

{"level":"error","ts":1696247731.0288842,"caller":"registry/registry.go:211","msg":"mirror failed attempting next","error":"expected mirror to respond with 200 OK but received: 500 Internal Server Error","stacktrace":"github.com/xenitab/spegel/internal/registry.(*Registry).handleMirror.func2\n\t/build/internal/registry/registry.go:211\nnet/http/httputil.(*ReverseProxy).modifyResponse\n\t/usr/local/go/src/net/http/httputil/reverseproxy.go:324\nnet/http/httputil.(*ReverseProxy).ServeHTTP\n\t/usr/local/go/src/net/http/httputil/reverseproxy.go:490\ngithub.com/xenitab/spegel/internal/registry.(*Registry).handleMirror\n\t/build/internal/registry/registry.go:217\ngithub.com/xenitab/spegel/internal/registry.(*Registry).registryHandler\n\t/build/internal/registry/registry.go:137\ngithub.com/gin-gonic/gin.(*Context).Next\n\t/go/pkg/mod/github.com/gin-gonic/gin@v1.9.1/context.go:174\ngithub.com/xenitab/spegel/internal/registry.(*Registry).metricsHandler\n\t/build/internal/registry/registry.go:271\ngithub.com/gin-gonic/gin.(*Context).Next\n\t/go/pkg/mod/github.com/gin-gonic/gin@v1.9.1/context.go:174\ngithub.com/gin-gonic/gin.CustomRecoveryWithWriter.func1\n\t/go/pkg/mod/github.com/gin-gonic/gin@v1.9.1/recovery.go:102\ngithub.com/gin-gonic/gin.(*Context).Next\n\t/go/pkg/mod/github.com/gin-gonic/gin@v1.9.1/context.go:174\ngithub.com/slok/go-http-metrics/middleware/gin.Handler.func1.1\n\t/go/pkg/mod/github.com/slok/go-http-metrics@v0.10.0/middleware/gin/gin.go:17\ngithub.com/slok/go-http-metrics/middleware.Middleware.Measure\n\t/go/pkg/mod/github.com/slok/go-http-metrics@v0.10.0/middleware/middleware.go:117\ngithub.com/slok/go-http-metrics/middleware/gin.Handler.func1\n\t/go/pkg/mod/github.com/slok/go-http-metrics@v0.10.0/middleware/gin/gin.go:16\ngithub.com/gin-gonic/gin.(*Context).Next\n\t/go/pkg/mod/github.com/gin-gonic/gin@v1.9.1/context.go:174\ngithub.com/xenitab/pkg/gin.Logger.func1\n\t/go/pkg/mod/github.com/xenitab/pkg/gin@v0.0.9/logger.go:28\ngithub.com/gin-gonic/gin.(*Context).Next\n\t/go/pkg/mod/github.com/gin-gonic/gin@v1.9.1/context.go:174\ngithub.com/gin-gonic/gin.(*Engine).handleHTTPRequest\n\t/go/pkg/mod/github.com/gin-gonic/gin@v1.9.1/gin.go:620\ngithub.com/gin-gonic/gin.(*Engine).ServeHTTP\n\t/go/pkg/mod/github.com/gin-gonic/gin@v1.9.1/gin.go:576\nnet/http.serverHandler.ServeHTTP\n\t/usr/local/go/src/net/http/server.go:2936\nnet/http.(*conn).serve\n\t/usr/local/go/src/net/http/server.go:1995"}
```

**Comment by infa-ddeore** [comment]
i see the caching isnt working at all, same image from different node took ~1m59s time always from different nodes, that means the image was downloaded from remote and not from local cache

**Comment by phillebaba** [comment]
Could you share your values.yaml used to deploy Spegel?

**Comment by phillebaba** [comment]
@infa-ddeore considering that the request that is failing is a HEAD and you are receiving a 500 my best guess is that something is happening when getting the size.
https://github.com/XenitAB/spegel/blob/8cf145f7cd0fb02b2a6be3ab13c255aae930745c/internal/registry/registry.go#L251-L256

To help you further I need the error message from the other Spegel instance. I have seen that OCI artifacts can differ a bit depending on how they are built, it may be that a specific field is missing from the artifact for example.

**Comment by infa-ddeore** [comment]
here is values.yaml, masked private repo locations
```
serviceMonitor:
  enabled: true

spegel:
  registries:
    - https://602401143452.dkr.ecr-fips.us-east-1.amazonaws.com
    - https://602401143452.dkr.ecr.us-west-2.amazonaws.com
    - https://containers.dl.xxx.io
    - https://xxx.artifacts.xxx.yyy
    - https://xxx.artifacts.aaa.yyy
    - https://docker.io
    - https://zzz.xxx.io
    - https://zzzz.xxx.io
    - https://qqq.xxx.io
    - https://qqqqq.xxx.io
    - https://qqqqqqq.xxx.io
    - https://qqq.xxx.io
    - https://qqqqqqq.xxx.io
    - https://k8s.gcr.io
    - https://public.ecr.aws
    - https://quay.io
    - https://registry.k8s.io
    - https://qqqqq.xxx.io
```
i downloaded `httpd` image from dockerhub with and without spegel, spegel is actually adding more delay in image download compared to direct

attaching `first.log` and `second.log` from the pods where image was pulled
1. image pull at first time took 1m3.800998679s
2. imagepull at second time from cache took 22.28909299s
3. imagepull directly without spegel took only 7.022743234s 

so spegel is adding more delay in image download instead of helping, all 404 have ~5 seconds time, looks like some timeout 

[first.log](https://github.com/XenitAB/spegel/files/12801074/first.log)
[second.log](https://github.com/XenitAB/spegel/files/12801076/second.log)

**Comment by infa-ddeore** [comment]
any pointers on the issue?

**Comment by phillebaba** [comment]
It is obvious now that I did not think about the default value for the lookup timeout, especially for images with lots of layers. I should have set a different default value that is a lot lower, which I am currently using myself. This is probably a good time for me to get going on performance tests to find an optimal timeout value. For now I will just change the default lookup timeout to something a lot lower.

**Comment by infa-ddeore** [comment]
ok, is that causing slowness in image download? non spegel cluster took 7 seconds for an image where spegel cluster took 1m3s without cache and with cache it was 22s, even with cache the download was slower, is it due to the timeout?

**Comment by tico24** [comment]
If it helps, I am experiencing the same issue. 

EKS 1.28. A mix of arm and amd64 nodes.

I thought the issue might be related to the mix of CPU architectures. I used a nodeselector to only use arm nodes, but got the same issue.

It's as if the spegel pods are not able to communicate with one another, but I can't work out what (if any) firewall/security group rules I may need to add.

I tried downgrading to spegel 0.0.11, but this didn't change things.

**Comment by tico24** [comment]
I downgraded AWS CNI from 0.15.x to 0.14.x and 0.13.x in the hopes that maybe the newere feature whereby it obeys `networkPolicies` might be causing the issue.

Unfortunately the issue still remains on 0.13.x

**Comment by tico24** [comment]
I downgraded AWS CNI from 0.15.x to 0.14.x and 0.13.x in the hopes that maybe the newere feature whereby it obeys networkPolicies might be causing the issue.

It **appears** to be working with AWS CNI 0.13.

**Comment by tico24** [comment]
Just to clarify, this speeds things up simply by timing out faster. But Spegel still isn't actually mirroring, right?

**Comment by phillebaba** [comment]
Well the issue that you are seeing is that layers that do not exist within the cluster will require the timeout to occur before allowing Containerd to pull from the original registry. This was introduced partly during some refactoring of the mirror component of the registry. I have a fix which would exit a lot faster when the layer cannot be found. Will try to get a new version out soon.

**Comment by tico24** [comment]
The layers definitely exist in the cluster. Logs are from the second pull (different node, same cluster)

**Comment by phillebaba** [comment]
I may have been looking at the wrong problem. The thing is that I am not seeing any major differences between my EKS and AKS deployments, but I am runnign Calico in both so it may be caused by the AWS CNI.

The issues that you are seeing may be two parts. One is that it cant find any peers due to network access issues. Another is that requests are waiting the max timeout when no peers are available, when in reality it should immediately fail because it has not peers.

The PR I created is meant to fix the latter, I will see if I can find an easy way to debug peer discovery issues.

**Comment by tico24** [comment]
In case it's useful, here's the exact steps I perform to test this.

1. Install spegel (via helm chart) and wait for everything to settle.
2. Run `kubectl run k8s-test --rm -i --tty --image ubuntu /bin/bash` This pulls the ubuntu image on a node and starts a pod.
3. When loaded, I corden the node that the pod has started on.
4. ctrl+d in the command line to kill the pod and wait for it to properly die.
5. Run `kubectl run k8s-test --rm -i --tty --image ubuntu /bin/bash` and observe the spegel logs


I also just tried the above with the following netpols, but no change. Still not mirroring.
```
apiVersion: networking.k8s.io/v1
kind: NetworkPolicy
metadata:
  name: allow-all-ingress
spec:
  podSelector: {}
  ingress:
  - {}
  policyTypes:
  - Ingress
---
apiVersion: networking.k8s.io/v1
kind: NetworkPolicy
metadata:
  name: allow-all-egress
spec:
  podSelector: {}
  egress:
  - {}
  policyTypes:
  - Egress
```

**Comment by infa-nang** [comment]
@phillebaba I'm also facing the same issue in EKS (v1.27).

But what I found was spegel was able to successfully mirror all non-blob layers, but when the mirror request comes-in for an blob layer spegel return 500 stating `content digest sha256 not found <sha256_value>` and this issue is because the containerd library spegel is using returns 404 for all blob layers.

I ssh'd into the EKS node and used the `ctr` cli, even with `ctr` I'm facing the same issue. I can able to get index/manifest/config layers using `./ctr content get sha256:<sha256_vaule_of_non_blob_layer>`, but when I try to get a blob layer with digest `ctr` returns not found.
I've used "quay.io/prometheus/alertmanager:v0.24.0" as reference, but the behavior is same will all registries.
* verifying "alertmanager" image is already present in the node 
![debug1](https://github.com/XenitAB/spegel/assets/113882397/e4a3bc0d-15aa-4886-acc2-c130c6ce3580)

* `regctl` cli output for list of layers present for "linux/amd" arch
![debug2](https://github.com/XenitAB/spegel/assets/113882397/f17256ac-0fff-439a-ad23-de5b73cd610e)

* I was able to get the content of the image's "manifest" type digest
![debug3](https://github.com/XenitAB/spegel/assets/113882397/6801571e-43c5-4878-8579-5fd6f7c7d1f0)

* When I try to retrieve a blob layer `ctr` returns not found, which the same behaviour we getting from containerd's golang package
![debug4](https://github.com/XenitAB/spegel/assets/113882397/10ca5998-567f-4b38-a39f-d14b12a941d7)

* But, if I use code with function similar to this https://github.com/XenitAB/spegel/blob/4a190529ade0eabbfcdc107c8173ef39b6c2f3b8/internal/oci/containerd.go#L150, I was able to get the list of all layers, but still not able to retrieve the layer data
![debug5](https://github.com/XenitAB/spegel/assets/113882397/39961308-bc98-41cf-9a51-521b8ba23c69).

If I use `ctr` to pull the image like `ctr image pull quay.io/quay/busybox:latest` then everything is working fine, I can able to retrieve the content of blob layers. But if I use an image which is already present in the machine (pulled via kubelet), then I can't able to get the blob data.

**Comment by infa-nang** [comment]
`discard_unpacked_layers = true` in the containerd config is the issue.
which repulls the blob layer every time making spegel slow

https://github.com/awslabs/amazon-eks-ami/blob/915ce2222e692a4d7b5904ab020d3c28e1e3e511/files/containerd-config.toml#L10

**Comment by tico24** [comment]
Just confirming that my world is much happier after tweaking the containerd config. If it's useful, here's how I did it using the [eks terraform module](https://registry.terraform.io/modules/terraform-aws-modules/eks/aws/latest):

```
      pre_bootstrap_user_data = <<-EOT
        #!/bin/bash
        cat /etc/containerd/config.toml | sed -E 's/discard_unpacked_layers = true/discard_unpacked_layers = false/g' > /etc/containerd/config-new.toml

        set -ex
        cat <<-EOF > /etc/profile.d/bootstrap.sh
        export CONTAINERD_CONFIG_FILE=/etc/containerd/config-new.toml
        EOF
        # Source extra environment variables in bootstrap script
        sed -i '/^set -o errexit/a\\nsource /etc/profile.d/bootstrap.sh' /etc/eks/bootstrap.sh
      EOT
```

**Comment by danielloader** [comment]
Going to have to port that logic over to a Karpenter AWSNodeProfile object at some point soon. Leaving this as a placeholder for when I work it out and edit this comment.

**Comment by phillebaba** [comment]
@infa-nang thanks for the great detective work determining the root cause of this issue! 

The kind cluster used for the e2e tests sets the same configuration.
https://github.com/XenitAB/spegel/blob/4a190529ade0eabbfcdc107c8173ef39b6c2f3b8/e2e/kind-config.yaml#L7-L11

There are now three main tasks that need to be done.
* Document the containerd configuration in the compatibility guide to help others.
* Implement a check and maybe log a warning if this is detected, alternatively exit.
* Create a performance test to document expected performance from Spegel.

**Comment by phillebaba** [comment]
It would be great if we could compile some examples for how to override the new default setting. That way end users do not have to go searching for a new solution every time.

@danielloader could you share how you solved this for Karpenter?

@tico24 in your example you create a new config file. Is this acutally required? I think that it would be nice to have an example that overrides the existing configuration file.

**Comment by tico24** [comment]
> @tico24 in your example you create a new config file. Is this acutally required? I think that it would be nice to have an example that overrides the existing configuration file.

Yup, it's required. The old configfile gets re-written to default after this script runs. Blame Jeff Bezos for that one.

**Comment by phillebaba** [comment]
Great, is it ok if I add it to the docs as a suggested solution?

**Comment by tico24** [comment]
Of course. At the moment it's quite specific to that terraform module. It can probably be adjusted to work for any Managed Node Group.

Ultiamtely, I believe you need the following in your managed nodegroup launch template config:
```
Content-Type: multipart/mixed; boundary="//"
MIME-Version: 1.0



--//

Content-Transfer-Encoding: 7bit

Content-Type: text/x-shellscript

Mime-Version: 1.0



#!/bin/bash
cat /etc/containerd/config.toml | sed -E 's/discard_unpacked_layers = true/discard_unpacked_layers = false/g' > /etc/containerd/config-new.toml

set -ex
cat <<-EOF > /etc/profile.d/bootstrap.sh
export CONTAINERD_CONFIG_FILE=/etc/containerd/config-new.toml
EOF
# Source extra environment variables in bootstrap script
sed -i '/^set -o errexit/a\\nsource /etc/profile.d/bootstrap.sh' /etc/eks/bootstrap.sh


--//--

```

This goes under `advanced details` > `user data`
![image](https://github.com/XenitAB/spegel/assets/45351296/3302ab0b-8475-4fbf-8ed5-a9aae59481ca)

**Comment by tico24** [comment]
The alternative would be to build custom AMIs.. but I started to go down that route and decided I'd rather not :)

**Comment by danielloader** [comment]
@phillebaba Updated above.
@tico24 I had to omit the newline character from the sed inline amend to get it to work, not sure if karpenter was mangling it when launching the instance but confirmed on the instance via SSM shell, that field had a prefix `n` character on the bootstrap.sh file.

**Comment by danielloader** [comment]
Additionally for historical context for people coming to this issue in the future: https://github.com/awslabs/amazon-eks-ami/issues/1359

**Comment by infa-ddeore** [comment]
thanks @phillebaba, so it just checks whether discardUnpackedLayers is set to true and gives error?

**Comment by phillebaba** [comment]
Yes the check is similar to the one for config path being set. Checking Containerd configuration, especially in managed Kuberentes, is cumbersome. It also requires the end user to know that the configuration needs to be checked. My guess is that 99% of users want Spegel to exit if the configuration is wrong.

**Comment by danielloader** [comment]
Update for the thread, I've had kubelet fail to start with the above patch on the latest 1.29 AMI - `amazon-eks-node-1.29-v20240129`.

Not sure why yet just thought I'd raise it as commenting out the userData patch got the nodes up and running.

**Comment by danielloader** [comment]
Redated log attached: 
[spegel-userdata.log](https://github.com/XenitAB/spegel/files/14164957/spegel-userdata.log)

**Comment by tico24** [comment]
This probably warrants a new issue (seeming as this one is closed).

I also have the same issue. I can run the commands manually on the node and everything seems fine. This is almost certainly an upstream aws issue, but good to raise for awareness.
